### ISSUE TITLE
belongsToMany's automatic setters delete data when passed a non-existing id

### ISSUE THREAD

**Issue by closingin** [sequelize/sequelize#13149]
## Issue Description

Hello there. I just came across a bug as I was trying to push invalid data to a junction table (a non-existing id from a table) using the automatic setters (`model.setFoo()`) from my `belongsToMany`  association. It deleted existing data before checking that it would be able to insert the new one.

For the usecase behind this, I have `events` for which you can upload `media` (photos, documents, ...), and assign them to an event through an `eventPhotos` junction table. The problem occurs when trying to assign a medium that has just been deleted for example.

### What are you doing?

Let's say my data is as follows :

```
SELECT id, size, path FROM "media"
----+---------+------------------------------------------------------------
 id |  size   |                            path                            
----+---------+------------------------------------------------------------
  1 |   35903 | /uploads/media/images/160145d02518dc23939bb548969b84de.png
  2 |  321789 | /uploads/media/images/8f4ed36200c8dff9637e3d7b9f42c298.png

SELECT id, name FROM "events"
----+--------------
 id |     name     
----+--------------
  1 | Sample event

SELECT * FROM "eventPhotos"
---------+----------
 eventId | mediumId 
---------+----------
       1 |        1
       1 |        2

```
```js
// Model setup

class Event extends Model {}
class Medium extends Model {}
class EventPhoto extends Model {}

Event.init({
  name: DataTypes.STRING
}, {})

Medium.init({
  size: DataTypes.INTEGER,
  path: DataTypes.STRING
}, {})


// Associations

EventPhoto.belongsTo(Event, { as: 'event', foreignKey: 'eventId' })
EventPhoto.belongsTo(Medium, { as: 'medium', foreignKey: 'mediumId' })
Event.belongsToMany(Medium, { as: 'photos', through: EventPhoto)


// Querying/Updating

const event = await Event.findOne(...)
await event.setPhotos([6])
```

### What do you expect to happen?

I think that the call should fail and the existing data should stay untouched.

### What is actually happening?

All data related to this event gets deleted from the junction table `eventPhotos`. I haven't checked the codebase behind this, but I would say that this is how it occurs:
- an array with new ids is received
- existing ids in the table don't exist anymore in the array, we can delete them
- insert rows with the ids from the array
- failure due to invalid id

I think it should first check that the received array contains valid data? That would be a "small" performance drop, but considering that this bug could lead to data loss, it may be worth the cost.

Also considering that automatic setters do the magic work of comparing existing data with the new data, I don't think that it's up to the user to sanitize ids.

### Environment

- Sequelize version: `6.6.2`
- Node.js version: `14.16.0`
- Operating System: Arch Linux

### How does this problem relate to dialects?

- [x] I think this problem happens regardless of the dialect.
- [ ] I think this problem happens only for the following dialect(s): 
- [ ] I don't know, I was using PUT-YOUR-DIALECT-HERE, with connector library version XXX and database version XXX

### Would you be willing to resolve this issue by submitting a Pull Request?

- [ ] Yes, I have the time and I know how to start.
- [x] Yes, I have the time but I don't know how to start, I would need guidance.
- [ ] No, I don't have the time, although I believe I could do it if I had the time...
- [ ] No, I don't have the time and I wouldn't even know how to start.

**Comment by jctaveras** [comment]
I agree with you that if you want to use the `.set` function Sequelize should not delete the previous entries. I can take a look at this and since I think I know what's happening.

**Comment by jctaveras** [comment]
@closingin I just opened [this PR](https://github.com/sequelize/sequelize/pull/13153) addressing your issue. Let me know if that is the expected behavior that you are looking for

**Comment by closingin** [comment]
Hello @jctaveras and thanks for your work on this issue!

I'll check it out asap, but expect it to be not until this week-end.

**Comment by closingin** [comment]
The issue is still open, we're waiting for the PR to be merged.
